File: src/main/kotlin/dev/murad/shipping/setup/ModMenuTypes.java
package dev.murad.shipping.setup;

import dev.murad.shipping.entity.accessor.EnergyHeadVehicleDataAccessor;
import dev.murad.shipping.entity.accessor.SteamHeadVehicleDataAccessor;
import dev.murad.shipping.entity.accessor.TugRouteScreenDataAccessor;
import dev.murad.shipping.entity.container.EnergyHeadVehicleContainer;
import dev.murad.shipping.entity.container.FishingBargeContainer;
import dev.murad.shipping.entity.container.SteamHeadVehicleContainer;
import dev.murad.shipping.entity.custom.train.locomotive.EnergyLocomotiveEntity;
import dev.murad.shipping.entity.custom.train.locomotive.SteamLocomotiveEntity;
import dev.murad.shipping.entity.custom.vessel.tug.EnergyTugEntity;
import dev.murad.shipping.entity.custom.vessel.tug.SteamTugEntity;
import dev.murad.shipping.item.container.TugRouteContainer;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.inventory.MenuType;
import net.minecraft.world.inventory.SimpleContainerData;
import net.neoforged.neoforge.common.extensions.IMenuTypeExtension;

import java.util.function.Supplier;


public class ModMenuTypes {

    private static SimpleContainerData makeIntArray(FriendlyByteBuf buffer) {
        int size = (buffer.readableBytes() + 1) / 4;
        SimpleContainerData arr = new SimpleContainerData(size);
        for (int i = 0; i < size; i++) {
            arr.set(i, buffer.readInt());
        }
        return arr;
    }

    public static final Supplier<MenuType<SteamHeadVehicleContainer<SteamTugEntity>>> TUG_CONTAINER =
            Registration.CONTAINERS.register("tug_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new SteamHeadVehicleContainer<>(windowId, inv.player.level(), new SteamHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player)));

    public static final Supplier<MenuType<EnergyHeadVehicleContainer<EnergyTugEntity>>> ENERGY_TUG_CONTAINER =
            Registration.CONTAINERS.register("energy_tug_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new EnergyHeadVehicleContainer<>(windowId, inv.player.level(), new EnergyHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player)));

    public static final Supplier<MenuType<SteamHeadVehicleContainer<SteamLocomotiveEntity>>> STEAM_LOCOMOTIVE_CONTAINER =
            Registration.CONTAINERS.register("steam_locomotive_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new SteamHeadVehicleContainer<>(windowId, inv.player.level(), new SteamHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player)));

    public static final Supplier<MenuType<EnergyHeadVehicleContainer<EnergyLocomotiveEntity>>> ENERGY_LOCOMOTIVE_CONTAINER =
            Registration.CONTAINERS.register("energy_locomotive_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new EnergyHeadVehicleContainer<>(windowId, inv.player.level(), new EnergyHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player)));

    public static final Supplier<MenuType<FishingBargeContainer>> FISHING_BARGE_CONTAINER =
            Registration.CONTAINERS.register("fishing_barge_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new FishingBargeContainer(windowId, inv.player.level(), data.readInt(), inv, inv.player)));

    public static final Supplier<MenuType<TugRouteContainer>> TUG_ROUTE_CONTAINER =
            Registration.CONTAINERS.register("tug_route_container",
                    () -> IMenuTypeExtension.create(
                            (windowId, inv, data) ->
                                    new TugRouteContainer(windowId, inv.player.level(), new TugRouteScreenDataAccessor(makeIntArray(data)), inv, inv.player)));




    public static void register () {}
}

File: src/main/kotlin/dev/murad/shipping/setup/ModMenuTypes.kt
package dev.murad.shipping.setup

import dev.murad.shipping.entity.accessor.EnergyHeadVehicleDataAccessor
import dev.murad.shipping.entity.accessor.SteamHeadVehicleDataAccessor
import dev.murad.shipping.entity.accessor.RouteScreenDataAccessor
import dev.murad.shipping.entity.container.EnergyHeadVehicleContainer
import dev.murad.shipping.entity.container.FishingBargeContainer
import dev.murad.shipping.entity.container.SteamHeadVehicleContainer
import dev.murad.shipping.entity.custom.train.locomotive.EnergyLocomotiveEntity
import dev.murad.shipping.entity.custom.train.locomotive.SteamLocomotiveEntity
import dev.murad.shipping.entity.custom.vessel.tug.EnergyTugEntity
import dev.murad.shipping.entity.custom.vessel.tug.SteamTugEntity
import dev.murad.shipping.item.container.RouteContainer
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.inventory.MenuType
import net.minecraft.world.inventory.SimpleContainerData
import net.neoforged.neoforge.common.extensions.IMenuTypeExtension
import java.util.function.Supplier

object ModMenuTypes {

    val TUG_CONTAINER: Supplier<MenuType<SteamHeadVehicleContainer<SteamTugEntity>>> =
        Registration.CONTAINERS.register("tug_container",
            Supplier {
                IMenuTypeExtension.create { windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf ->
                    SteamHeadVehicleContainer(
                        windowId, inv.player.level(), SteamHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player
                    )
                }
            })

    val ENERGY_TUG_CONTAINER: Supplier<MenuType<EnergyHeadVehicleContainer<EnergyTugEntity>>> =
        Registration.CONTAINERS.register("energy_tug_container",
            Supplier {
                IMenuTypeExtension.create { windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf ->
                    EnergyHeadVehicleContainer(
                        windowId, inv.player.level(), EnergyHeadVehicleDataAccessor(makeIntArray(data)), inv, inv.player
                    )
                }
            })

    val STEAM_LOCOMOTIVE_CONTAINER: Supplier<MenuType<SteamHeadVehicleContainer<SteamLocomotiveEntity>>> =
        Registration.CONTAINERS.register("steam_locomotive_container",
            Supplier{IMenuTypeExtension.create(::steamHeadVehicleContainer)}
        )


    private fun steamHeadVehicleContainer(windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf):
            SteamHeadVehicleContainer<SteamLocomotiveEntity> {
        return SteamHeadVehicleContainer(
            windowId,
            inv.player.level(),
            SteamHeadVehicleDataAccessor(makeIntArray(data)),
            inv,
            inv.player
        )
    }

    val ENERGY_LOCOMOTIVE_CONTAINER: Supplier<MenuType<EnergyHeadVehicleContainer<EnergyLocomotiveEntity>>> =
        Registration.CONTAINERS.register("energy_locomotive_container",
            Supplier {
                IMenuTypeExtension.create { windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf ->
                    EnergyHeadVehicleContainer(
                        windowId, inv.player.level(), EnergyHeadVehicleDataAccessor(
                            makeIntArray(data)
                        ), inv, inv.player
                    )
                }
            })

    val FISHING_BARGE_CONTAINER: Supplier<MenuType<FishingBargeContainer>> =
        Registration.CONTAINERS.register("fishing_barge_container",
            Supplier {
                IMenuTypeExtension.create { windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf ->
                    FishingBargeContainer(
                        windowId,
                        inv.player.level(),
                        data?.readInt(),
                        inv,
                        inv.player
                    )
                }
            })


    val TUG_ROUTE_CONTAINER: Supplier<MenuType<RouteContainer>> =
        Registration.CONTAINERS.register("tug_route_container",
            Supplier {
                IMenuTypeExtension.create { windowId: Int, inv: Inventory, data: RegistryFriendlyByteBuf ->
                    RouteContainer(
                        windowId, RouteScreenDataAccessor(
                            makeIntArray(data)
                        ), inv.player
                    )
                }
            })


    fun register() {}

    //on a NPE check if somewhere player.openMenu(...) is called without 2nd param  getDataAccessor(...)::write
    private fun makeIntArray(buffer: FriendlyByteBuf): SimpleContainerData {

        val size = (buffer.readableBytes() + 1) / 4
        val arr = SimpleContainerData(size)
        for (i in 0 until size) {
            arr[i] = buffer.readInt()
        }
        return arr
    }
}
